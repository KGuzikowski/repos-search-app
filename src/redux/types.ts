import { ThunkAction } from 'redux-thunk'
import { Action } from 'redux'

export type AppThunk<State, ReturnType = void> = ThunkAction<ReturnType, State, unknown, Action<string>>