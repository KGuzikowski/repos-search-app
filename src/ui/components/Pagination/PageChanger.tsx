import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { RootState } from '../../../redux/rootReducer'
import { setPage } from '../../../redux/repos/repo.actions'
import { PageChangerContainer, Option, OptionsDiv, Of, Dots } from './Pagination.styles'
import { ArrowRight, ArrowLeft } from '../TableSection/Arrows'

interface PageChangerType {
    reposNum: number,
    start: number,
    end: number
}

const PageChanger = ({ reposNum, start, end }: PageChangerType) => {
    const dispatch = useDispatch()
    const perPage = useSelector((state: RootState) => state.repos.pagination)
    const currentPage = useSelector((state: RootState) => state.repos.page)

    const handleOptionClick = (e: React.MouseEvent<HTMLParagraphElement>) => {
        const nextPage = parseInt(e.currentTarget.innerHTML)
        if(currentPage !== nextPage)
            dispatch(setPage(nextPage))
    }

    const handleArrowClick = (num: number) => {
        dispatch(setPage(currentPage + num))
    }

    const pagesNumber = Math.ceil(reposNum / perPage)
    let nextPages: number[] = []
    
    if(currentPage === 1)
        for(let i = 2; i < currentPage + 4 && i < pagesNumber; i++) nextPages.push(i)
    else if(currentPage === pagesNumber - 2) {
        nextPages.push(currentPage-1)
        nextPages.push(currentPage)
        nextPages.push(currentPage+1)
    } else if(currentPage === pagesNumber - 1) {
        nextPages.push(currentPage-2)
        nextPages.push(currentPage-1)
        nextPages.push(currentPage)
    } else if(currentPage === pagesNumber) {
        nextPages.push(currentPage-3)
        nextPages.push(currentPage-2)
        nextPages.push(currentPage-1)
    }
    else for(let i = currentPage; i < currentPage + 3 && i < pagesNumber; i++) nextPages.push(i)

    return (
        <PageChangerContainer>
            <div>{start + 1}-{ end > reposNum ? reposNum : end } of {reposNum} repos</div>
            <OptionsDiv>
                <ArrowLeft onClick={() => handleArrowClick(-1)}/>
                <Option current={1 === currentPage} onClick={(e) => handleOptionClick(e)}>
                    1
                </Option>
                {currentPage > 2 ? <Dots>...</Dots> : null}
                {nextPages.map(num => <Option key={num} current={num === currentPage} onClick={(e) => handleOptionClick(e)}>{num}</Option>)}
                <Of>of</Of>
                <Option current={currentPage === pagesNumber} onClick={(e) => handleOptionClick(e)}>
                    {pagesNumber}
                </Option>
                <ArrowRight onClick={() => handleArrowClick(1)}/>
            </OptionsDiv>
        </PageChangerContainer>
    )
}

export default PageChanger