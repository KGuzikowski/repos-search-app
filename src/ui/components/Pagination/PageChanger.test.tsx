import React from 'react'
import { create } from 'react-test-renderer'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import PageChanger from './PageChanger'
import { reposStateType } from '../../../redux/repos/repo.types'

it('renders PageChanger component correctly', () => {
    const repo = {
        id: 1,
        name: 'repo',
        owner: { id: 1, name: 'Karol' },
        stars: 2,
        createdAt: '2019-05-12'
    }

    const props = {
        reposNum: 1,
        start: 1,
        end: 1
    }

    const INITIAL_STATE: reposStateType = {
        repos: [repo],
        isFetching: false,
        errorMessage: '',
        name: 'react',
        pagination: 5,
        page: 1,
    }

    const mockStore = configureStore()
    const store = mockStore(INITIAL_STATE)

    const component = create(<Provider store={store}><PageChanger {...props} /></Provider>)
    expect(component.toJSON()).toMatchSnapshot()
})