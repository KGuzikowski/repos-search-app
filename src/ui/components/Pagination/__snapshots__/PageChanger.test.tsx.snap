// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PageChanger component correctly 1`] = `
<div
  className="sc-AxhCb eqnjqy"
>
  <div>
    2
    -
    1
     of 
    1
     repos
  </div>
  <div
    className="sc-Axmtr gxCDUi"
  >
    <svg
      className="sc-fzpans hKpuCT"
      onClick={[Function]}
    >
      arrow_down.svg
    </svg>
    <p
      className="sc-AxhUy goOHvP"
      onClick={[Function]}
    >
      1
    </p>
    <p
      className="sc-AxheI jzDweV"
    >
      of
    </p>
    <p
      className="sc-AxhUy goOHvP"
      onClick={[Function]}
    >
      NaN
    </p>
    <svg
      className="sc-fzoLsD dnFGMq"
      onClick={[Function]}
    >
      arrow_down.svg
    </svg>
  </div>
</div>
`;
